src/pk.c

```diff
@@ -51,6 +51,487 @@ static int pem_write_bio_pubkey(WOLFSSL_BIO* bio, WOLFSSL_EVP_PKEY* key);
+                if (p == NULL) {
+                    sz = MEMORY_E;
+                    break;
+                }
+
+                /* Set mem to new pointer. */
+                mem = p;
+                /* Set p to where to read in next chunk. */
+                p += ret;
+                /* Read in a new chunk. */
+                sz = READ_BIO_FILE_CHUNK;
+            }
+        }
+        if ((sz < 0) || (ret == 0)) {
+            /* Dispose of memory on error or no data read. */
+            XFREE(mem, bio->heap, DYNAMIC_TYPE_OPENSSL);
+            mem = NULL;
+            /* Return error. */
+            ret = sz;
+        }
+    }
+
+    *pem = mem;
+    return ret;
+}
+#endif
+
+/* Read exactly the required amount into a newly allocated buffer.
+ *
+ * @param [in, out] bio  BIO object to read with.
+ * @parma [in       sz   Amount of data to read.
+ * @parma [out]     pem  Read data in a new buffer.
+ * @return  -ve on error.
+ * @return  Number of bytes read on success.
+ */
+static int wolfssl_read_bio_len(WOLFSSL_BIO* bio, int sz, char** pem)
+{
+    int ret = 0;
+    char* mem;
+
+    /* Allocate buffer to hold data. */
+    mem = (char*)XMALLOC(sz, bio->heap, DYNAMIC_TYPE_OPENSSL);
+    if (mem == NULL) {
+        WOLFSSL_MSG("Memory error");
+        ret = MEMORY_E;
+    }
+    else if ((ret = wolfSSL_BIO_read(bio, mem, sz)) != sz) {
+        /* Pending data not read. */
+        XFREE(mem, bio->heap, DYNAMIC_TYPE_OPENSSL);
+        mem = NULL;
+        ret = MEMORY_E;
+    }
+
+    *pem = mem;
+    return ret;
+}
+
+/* Read all the data from a BIO.
+ *
+ * @param [in, out] bio         BIO object to read with.
+ * @param [out]     pem         Read data in a buffer.
+ * @param [out]     pemSz       Amount of data read in bytes.
+ * @param [out]     memAlloced  Indicates whether return buffer was allocated.
+ * @return  -ve on error.
+ * @return  0 on success.
+ */
+static int wolfssl_read_bio(WOLFSSL_BIO* bio, char** pem, int* pemSz,
+    int* memAlloced)
+{
+    int ret;
+    int sz;
+
+    if (bio->type == WOLFSSL_BIO_MEMORY) {
+        ret = wolfSSL_BIO_get_mem_data(bio, pem);
+        if (ret > 0) {
+            bio->rdIdx += ret;
+        }
+        *memAlloced = 0;
+    }
+#ifndef WOLFSSL_NO_FSEEK
+    /* Get pending or, when a file BIO, get length of file. */
+    else if ((sz = wolfSSL_BIO_get_len(bio)) > 0) {
+        ret = wolfssl_read_bio_len(bio, sz, pem);
+        if (ret > 0) {
+            *memAlloced = 1;
+        }
+    }
+#else
+    else if ((sz = wolfSSL_BIO_pending(bio)) > 0) {
+        ret = wolfssl_read_bio_len(bio, sz, pem);
+        if (ret > 0) {
+            *memAlloced = 1;
+        }
+    }
+    else if (bio->type == WOLFSSL_BIO_FILE) {
+        ret = wolfssl_read_bio_file(bio, pem);
+        if (ret > 0) {
+            *memAlloced = 1;
+        }
+    }
+#endif
+    else {
+        WOLFSSL_MSG("No data to read from bio");
+        *memAlloced = 0;
+        ret = NOT_COMPILED_IN;
```

Comment:
Seems like an odd return code for this error.

Reply:
Error is that we used an unsupported BIO type.
Not supported because it wasn't compiled in.